src/messages/mod.rs

```diff
@@ -15,29 +18,100 @@ mod test_proposals;
 #[cfg(test)]
 mod test_welcome;
 
+/// Welcome Messages
+///
+/// > 11.2.2. Welcoming New Members
+///
+/// ```text
+/// struct {
+///   ProtocolVersion version = mls10;
+///   CipherSuite cipher_suite;
+///   EncryptedGroupSecrets secrets<0..2^32-1>;
+///   opaque encrypted_group_info<1..2^32-1>;
+/// } Welcome;
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct Welcome {
+    version: ProtocolVersion,
+    cipher_suite: &'static Ciphersuite,
+    secrets: Vec<EncryptedGroupSecrets>,
+    encrypted_group_info: Vec<u8>,
+}
+
+/// EncryptedGroupSecrets
+///
+/// > 11.2.2. Welcoming New Members
+///
+/// ```text
+/// struct {
+///   opaque key_package_hash<1..255>;
+///   HPKECiphertext encrypted_group_secrets;
+/// } EncryptedGroupSecrets;
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct EncryptedGroupSecrets {
+    pub key_package_hash: Vec<u8>,
+    pub encrypted_group_secrets: HpkeCiphertext,
+}
+
+impl Welcome {
+    /// Create a new welcome message from the provided data.
+    /// Note that secrets and the encrypted group info are consumed.
+    pub(crate) fn new(
+        version: ProtocolVersion,
+        cipher_suite: &'static Ciphersuite,
+        secrets: Vec<EncryptedGroupSecrets>,
+        encrypted_group_info: Vec<u8>,
+    ) -> Self {
+        Self {
+            version,
+            cipher_suite,
+            secrets,
+            encrypted_group_info,
+        }
+    }
+
+    /// Get a reference to the ciphersuite in this Welcome message.
+    pub(crate) fn ciphersuite(&self) -> &'static Ciphersuite {
+        self.cipher_suite
+    }
+
+    /// Get a reference to the ciphersuite in this Welcome message.
+    pub fn get_secrets_ref(&self) -> &[EncryptedGroupSecrets] {
+        &self.secrets
+    }
+
+    /// Get a reference to the encrypted group info.
+    pub(crate) fn get_encrypted_group_info_ref(&self) -> &[u8] {
+        &self.encrypted_group_info
+    }
+}
+
 #[derive(Debug, PartialEq, Clone)]
 pub struct Commit {
-    pub proposals: Vec<ProposalID>,
-    pub path: Option<UpdatePath>,
+    pub(crate) proposals: Vec<ProposalID>,
+    pub(crate) path: Option<UpdatePath>,
 }
 
-impl Codec for Commit {
-    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
-        encode_vec(VecSize::VecU32, buffer, &self.proposals)?;
-        self.path.encode(buffer)?;
-        Ok(())
-    }
-    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
-        let proposals = decode_vec(VecSize::VecU32, cursor)?;
-        let path = Option::<UpdatePath>::decode(cursor)?;
-        Ok(Commit { proposals, path })
+impl Commit {
+    /// Returns `true` if the commit contains an update path. `false` otherwise.
+    pub fn has_path(&self) -> bool {
```

Comment:
Flag this as `#[cfg(test)]`? It doesn't seem to be used elsewhere. Or do we want this to be public for some other reason?

Reply:
Yeah the `Commit` struct itself is a little weird. This function only makes it possible to remove the `pub` on `path`. The entire thing needs to be rewritten.
Unfortunately `#[cfg(test)]` doesn't work because this is the library and that's not built with `-cfg test`.

Reply:
#236